sql/hive/src/test/scala/org/apache/spark/sql/hive/client/VersionsSuite.scala

```diff
@@ -862,17 +859,17 @@ class VersionsSuite extends SparkFunSuite with Logging {
             |        "logicalType": "decimal"
             |      }
             |    ]
-            |  } ]
+            |  }]
             |}
           """.stripMargin
-        val schemaUrl = s"""$schemaPath${File.separator}avroDecimal.avsc"""
-        val schemaFile = new File(schemaPath, "avroDecimal.avsc")
+        val schemaFile = new File(dir, "avroDecimal.avsc")
         val writer = new PrintWriter(schemaFile)
         writer.write(avroSchema)
         writer.close()
+        val schemaPath = schemaFile.getCanonicalPath
```

Comment:
Could we be able to just use `toURI` instead of `getCanonicalPath`?

Reply:
what's the difference? we use `getCanonicalPath` in a lot of tests...

Reply:
Ah, just for the test on Windows. Windows paths are like `C:\a\b\c` does not look going though the parser as is. It looks requiring escaping it .. Probably the easiest way is using URI to easily make the tests pass on Windows and Linux.

I have fixed those tests to pass on Windows few times. Up to my knowledge, there is a weak promise that Spark can run on Windows - https://spark.apache.org/docs/latest/#downloading. IMHO, it might be a bit better to use URI always when we remember this, and if those tests are not specific to path forms ..

Reply:
I just ran a test on Windows to double check:

Build started: [SQL] `org.apache.spark.sql.hive.client.VersionsSuite` [![PR-19799](https://ci.appveyor.com/api/projects/status/github/spark-test/spark?branch=E6B8B497-CBC7-4978-A985-79302759F300&svg=true)](https://ci.appveyor.com/project/spark-test/spark/branch/E6B8B497-CBC7-4978-A985-79302759F300)
Diff: https://github.com/apache/spark/compare/master...spark-test:E6B8B497-CBC7-4978-A985-79302759F300

Reply:
how does `toURI` help here? it will escape the separator for us?

Reply:
For example, `/` in `file:/C:/a/b/c` is fine but problem seems about `\` in `C:\a\b\c`. I believe there are few issues open about this.

For example, on Windows:

```

scala> sql("select '\\\\'").show()
17/11/23 05:04:37 WARN SizeEstimator: Failed to check whether UseCompressedOops
is set; assuming yes
+---+
|  \|
+---+
|  \|
+---+


scala> sql("select '\\'").show()
org.apache.spark.sql.catalyst.parser.ParseException:
no viable alternative at input 'select ''(line 1, pos 7)

== SQL ==
select '\'
-------^^^

  at org.apache.spark.sql.catalyst.parser.ParseException.withCommand(ParseDriver
.scala:239)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parse(ParseDriver.sc
ala:115)
  at org.apache.spark.sql.execution.SparkSqlParser.parse(SparkSqlParser.scala:48
)
  at org.apache.spark.sql.catalyst.parser.AbstractSqlParser.parsePlan(ParseDrive
r.scala:69)
  at org.apache.spark.sql.SparkSession.sql(SparkSession.scala:638)
  ... 49 elided

scala> sql("select '/'").show()
+---+
|  /|
+---+
|  /|
+---+

scala> sql("select '//'").show()
+---+
| //|
+---+
| //|
+---+
```

So ... I was thinking using `toURI`

```scala
scala> new File("C:\\a\\b\\c").toURI.toString
res6: String = file:/C:/a/b/c
```

Reply:
Sorry I don't have a windowns machine to test, do you mean `new File("C:\\a\\b\\c").toURI.toString` is different from `new File("C:\\a\\b\\c").getCanonicalPath` excluding the `file:` prefix?

Reply:
Sure, feel free to ask to test in the future as well. I can easily test.

Yes .. here:

```scala
scala> val path = new File("C:\\a\\b\\c").getCanonicalPath
path: String = C:\a\b\c

scala> sql(s"SELECT '$path'").show()
+-----+
|C:c|
+-----+
|C:c|
+-----+

scala> val path = new File("C:\\a\\b\\c").toURI.toString
path: String = file:/C:/a/b/c

scala> sql(s"SELECT '$path'").show()
+--------------+
|file:/C:/a/b/c|
+--------------+
|file:/C:/a/b/c|
+--------------+
```

Reply:
Leaving it as is is okay to me actually. I can fix them in a batch with more descriptions and explanation next time. We don't run Scala tests on Windows anyway as a requirement.

Reply:
ah good to know this. Let's do a holistic review and fix all of them at once.